include/swift/Sema/IDETypeChecking.h

```diff
@@ -40,6 +40,10 @@ namespace swift {
   /// \returns true on convertible, false on not.
   bool isConvertibleTo(Type T1, Type T2, DeclContext &DC);
 
+  /// \brief Determine whether a declaration is implicitly convertible to a
+  /// given type. Unlike isConvertibleTo, this handles generic types correctly.
+  bool isDeclConvertibleTo(const ValueDecl *decl, Type fromType, Type toType);
```

Comment:
If we have a `self` type, `isDeclConvertibleTo` could add a constraint that binds the 'self' type of a method to a specific type. That binding lets us provide types we know from the caller. For something like:

    func f<T>(_ fn: (T) -> Bool) { }

    struct X<U> {
      func g(_ fn: (U) -> Bool) { }
    }

    let xi = X<Int>()
    f(xi.<#code completion here#>)

At the code completion point, when checking where "xi.g" can be converted to the type expected by "f", we'd like to substitute the type of `xi` (`X<Int>`) through the type of `g`. The interface type of `g` is something like:

    <U> (self: X<U>) -> ((U) -> Bool) -> Void

If we just drop the first parameter, we end up with

  <U> ((U) -> Bool) -> Void

but if we substitute the 'self' type instead, we get a much more useful type:

  ((Int) -> Bool) -> Void

`ConstraintSystem::getTypeOfMemberReference()` is good at doing this substitution.